sdk/ai/azure-ai-inference/src/main/java/com/azure/ai/inference/ChatCompletionClientTracer.java

```diff
@@ -0,0 +1,606 @@
+        }
+        if (request.getTemperature() != null) {
+            tracer.setAttribute("gen_ai.request.temperature", request.getTemperature(), span);
+        }
+        if (request.getTopP() != null) {
+            tracer.setAttribute("gen_ai.request.top_p", request.getTopP(), span);
+        }
+        if (host != null) {
+            tracer.setAttribute("server.address", host, span);
+            tracer.setAttribute("server.port", port, span);
+        }
+    }
+
+    private void traceCompletionRequestEvents(List<ChatRequestMessage> messages, Context span) {
+        if (!captureContent) {
+            return;
+        }
+        if (messages != null) {
+            for (ChatRequestMessage message : messages) {
+                final ChatRole role = message.getRole();
+                if (role != null) {
+                    final String eventName = "gen_ai." + role.getValue() + ".message";
+                    final String eventContent = toJsonString(message);
+                    if (eventContent != null) {
+                        final Map<String, Object> eventAttributes = new HashMap<>(2);
+                        eventAttributes.put("gen_ai.system", INFERENCE_GEN_AI_SYSTEM_NAME);
+                        eventAttributes.put("gen_ai.event.content", eventContent);
+                        tracer.addEvent(eventName, eventAttributes, OffsetDateTime.now(ZoneOffset.UTC), span);
+                    }
+                }
+            }
+        }
+    }
+
+    private void traceCompletionResponseAttributes(ChatCompletions response, Context span) {
+        tracer.setAttribute("gen_ai.response.id", response.getId(), span);
+        tracer.setAttribute("gen_ai.response.model", response.getModel(), span);
+        final CompletionsUsage usage = response.getUsage();
+        if (usage != null) {
+            tracer.setAttribute("gen_ai.usage.input_tokens", usage.getPromptTokens(), span);
+            tracer.setAttribute("gen_ai.usage.output_tokens", usage.getCompletionTokens(), span);
+        }
+        final List<ChatChoice> choices = response.getChoices();
+        if (choices != null) {
+            tracer.setAttribute("gen_ai.response.finish_reasons", getFinishReasons(choices), span);
+        }
+    }
+
+    private void traceCompletionResponseAttributes(StreamingChatCompletionsUpdate response, String finishReasons,
+        Context span) {
+        tracer.setAttribute("gen_ai.response.id", response.getId(), span);
+        tracer.setAttribute("gen_ai.response.model", response.getModel(), span);
+        final CompletionsUsage usage = response.getUsage();
+        if (usage != null) {
+            tracer.setAttribute("gen_ai.usage.input_tokens", usage.getPromptTokens(), span);
+            tracer.setAttribute("gen_ai.usage.output_tokens", usage.getCompletionTokens(), span);
+        }
+        tracer.setAttribute("gen_ai.response.finish_reasons", finishReasons, span);
```

Comment:
I don't think we support array attributes in otel implementation - we should just add fix it in 

https://github.com/Azure/azure-sdk-for-java/blob/25dfa3de5300583a3986f29f9626efc640434230/sdk/core/azure-core-tracing-opentelemetry/src/main/java/com/azure/core/tracing/opentelemetry/OpenTelemetryUtils.java#L78

UPDATE: I realized you string-join them - no need, otel supports array (not list) attributes, we can just update our utils

Reply:
Hi @lmolkova thank you, I might be missing something, but I do not see a method in the [OTEL Span type](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/trace/Span.java) that takes "a key" with "value as array.". I see an [API ](https://github.com/open-telemetry/opentelemetry-java/blob/b32f64420c84bd2af7e60261baad01a198162639/api/all/src/main/java/io/opentelemetry/api/trace/Span.java#L182) that takes an "Attributes" instance, which can be built by adding key-value pairs to an "AttributesBuilder". However, this approach requires each key to be unique; otherwise, it will be overwritten. Our use case involves associating one key with multiple values. Maybe I am not looking in the right place. :sweat_smile:

Reply:
you can do something like this 

`span.setAttribute(AttributeKey.stringArrayKey("...finish_reasons"), Arrays.asList( "stop", "length" ))`

(it'd take care of list -> array conversion over the wire)

Reply:
Thank you @lmolkova for the pointers! My understanding is, you're proposing we add method(s) to ` om.azure.core.util.tracing.Tracer` in "azure-core" to support attributes values list, right? Is the proposal here to add following methods in `Tracer`

```java
Public interface Tracer {
    void setStringAttributes(String key, List<String> values, Context context);
    void setBooleanAttributes(String key, List<Boolean> values, Context context);
    void setDoubleAttributes(String key, List<Double> values, Context context);
    void setLongAttributes(String key, List<Long> values, Context context);
}
```

Reply:
this looks great, I'd prefer one method though

`void setLongAttributes(String key, Object, Context context)` if we can recover erased type. It'd end up in this switch as `Object` anyway

https://github.com/Azure/azure-sdk-for-java/blob/b8a9d8c1da8114ed9ed1c451a0549f56d4d9b510/sdk/core/azure-core-tracing-opentelemetry/src/main/java/com/azure/core/tracing/opentelemetry/OpenTelemetryUtils.java#L110

Reply:
We might need to do it as a follow up though.

@alzimmermsft are we planning to ship core in Jan or we're already in the quarterly release era?

Reply:
Agreed, let's handle it in a separate work item. Update to use that API in azure-ai-inference
can be a follow up pr. Created the work item. https://github.com/Azure/azure-sdk-for-java/issues/43408